File: backend/addons/quelyos_ecommerce/controllers/cms.py
# -*- coding: utf-8 -*-

from odoo import http, fields
from odoo.http import request
from .base_controller import BaseEcommerceController
import logging

_logger = logging.getLogger(__name__)


class CmsMenuController(BaseEcommerceController):
    """API Controller pour les menus CMS."""

    @http.route('/api/ecommerce/menus', type='json', auth='public', methods=['POST'], csrf=False, cors='*')
    def get_menus(self, codes=None, **kwargs):
        """
        Récupère les menus du site.

        Args:
            codes: Liste des codes de menus (ex: ['header', 'footer_quick'])
                   Si None, retourne tous les menus actifs

        Returns:
            {success: True, menus: {code: {...}, ...}}
        """
        try:
            domain = [('active', '=', True)]
            if codes:
                domain.append(('code', 'in', codes))

            menus = request.env['cms.menu'].sudo().search(domain)

            result = {}
            for menu in menus:
                result[menu.code] = menu.get_menu_data()

            return self._success_response({'menus': result})

        except Exception as e:
            return self._handle_error(e, "récupération des menus")

    @http.route('/api/ecommerce/menus/<string:code>', type='json', auth='public', methods=['POST'], csrf=False, cors='*')
    def get_menu(self, code, **kwargs):
        """
        Récupère un menu spécifique par son code.

        Args:
            code: Code technique du menu (ex: 'header')

        Returns:
            {success: True, menu: {...}}
        """
        try:
            menu = request.env['cms.menu'].sudo().search([
                ('code', '=', code),
                ('active', '=', True)
            ], limit=1)

            if not menu:
                return {
                    'success': False,
                    'error': f'Menu "{code}" non trouvé'
                }

            return self._success_response({'menu': menu.get_menu_data()})

        except Exception as e:
            return self._handle_error(e, f"récupération du menu {code}")


class CmsPageController(BaseEcommerceController):
    """API Controller pour les pages CMS."""

    @http.route('/api/ecommerce/pages', type='json', auth='public', methods=['POST'], csrf=False, cors='*')
    def get_pages(self, limit=20, offset=0, parent_id=None, template=None, **kwargs):
        """
        Liste les pages publiées.

        Args:
            limit: Nombre de pages (max 50)
            offset: Offset de pagination
            parent_id: ID de la page parente (optionnel)
            template: Filtrer par template (optionnel)

        Returns:
            {success: True, pages: [...], total: int}
        """
        try:
            now = fields.Datetime.now()
            domain = [
                ('state', '=', 'published'),
                ('active', '=', True),
                '|', ('publish_date', '=', False), ('publish_date', '<=', now),
                '|', ('unpublish_date', '=', False), ('unpublish_date', '>', now),
            ]

            if parent_id:
                domain.append(('parent_id', '=', int(parent_id)))

            if template:
                domain.append(('template', '=', template))

            Page = request.env['cms.page'].sudo()
            total = Page.search_count(domain)

            # Limiter à 50 max
            limit = min(int(limit or 20), 50)
            offset = int(offset or 0)

            pages = Page.search(domain, limit=limit, offset=offset, order='sequence, name')

            return self._success_response({
                'pages': [p.get_page_summary() for p in pages],
                'total': total,
                'limit': limit,
                'offset': offset
            })

        except Exception as e:
            return self._handle_error(e, "récupération des pages")

    @http.route('/api/ecommerce/pages/<string:slug>', type='json', auth='public', methods=['POST'], csrf=False, cors='*')
    def get_page(self, slug, **kwargs):
        """
        Récupère une page par son slug.

        Args:
            slug: Slug URL de la page (ex: 'about')

        Returns:
            {success: True, page: {...}}
        """
        try:
            page = request.env['cms.page'].sudo().get_published_page(slug)

            if not page:
                return {
                    'success': False,
                    'error': f'Page "{slug}" non trouvée'
                }

            # Incrémenter le compteur de vues
            page.increment_view_count()

            return self._success_response({'page': page.get_page_data()})

        except Exception as e:
            return self._handle_error(e, f"récupération de la page {slug}")

    @http.route('/api/ecommerce/pages/<string:slug>/seo', type='json', auth='public', methods=['POST'], csrf=False, cors='*')
    def get_page_seo(self, slug, **kwargs):
        """
        Récupère uniquement les données SEO d'une page.
        Utile pour le pre-rendering des meta tags côté frontend.

        Args:
            slug: Slug URL de la page

        Returns:
            {success: True, seo: {...}}
        """
        try:
            page = request.env['cms.page'].sudo().get_published_page(slug)

            if not page:
                return {
                    'success': False,
                    'error': f'Page "{slug}" non trouvée'
                }

            return self._success_response({'seo': page._get_seo_data()})

        except Exception as e:
            return self._handle_error(e, f"récupération SEO de {slug}")


class CmsBlockController(BaseEcommerceController):
    """API Controller pour les blocs CMS réutilisables."""

    @http.route('/api/ecommerce/blocks/<string:code>', type='json', auth='public', methods=['POST'], csrf=False, cors='*')
    def get_block(self, code, **kwargs):
        """
        Récupère un bloc réutilisable par son code.

        Args:
            code: Code technique du bloc

        Returns:
            {success: True, block: {...}}
        """
        try:
            block = request.env['cms.block'].sudo().search([
                ('code', '=', code),
                ('active', '=', True)
            ], limit=1)

            if not block:
                return {
                    'success': False,
                    'error': f'Bloc "{code}" non trouvé'
                }

            return self._success_response({'block': block.get_block_data()})

        except Exception as e:
            return self._handle_error(e, f"récupération du bloc {code}")

    @http.route('/api/ecommerce/blocks', type='json', auth='public', methods=['POST'], csrf=False, cors='*')
    def get_blocks(self, codes=None, **kwargs):
        """
        Récupère plusieurs blocs par leurs codes.

        Args:
            codes: Liste des codes de blocs à récupérer

        Returns:
            {success: True, blocks: {code: {...}, ...}}
        """
        try:
            domain = [('active', '=', True)]
            if codes:
                domain.append(('code', 'in', codes))

            blocks = request.env['cms.block'].sudo().search(domain)

            result = {block.code: block.get_block_data() for block in blocks}

            return self._success_response({'blocks': result})

        except Exception as e:
            return self._handle_error(e, "récupération des blocs")

File: vitrine-client/src/lib/odoo/cms.ts
/**
 * Service CMS pour récupérer les menus, pages et blocs depuis Odoo
 */

import axios from 'axios';
import type {
  Menu,
  CmsPage,
  CmsPageSummary,
  PageSeo,
  CmsBlock,
  GetPagesOptions,
  PagesResponse,
} from '@/types/cms';
import { logger } from '@/lib/logger';

// Use absolute URL for SSR, relative for client-side
const getApiBase = () => {
  if (typeof window === 'undefined') {
    // Server-side: use absolute URL
    const siteUrl = process.env.NEXT_PUBLIC_SITE_URL || 'http://localhost:3000';
    return `${siteUrl}/api/odoo`;
  }
  // Client-side: use relative URL
  return '/api/odoo';
};

/**
 * Appel JSON-RPC vers l'API CMS
 */
async function jsonrpc<T = any>(
  endpoint: string,
  params: Record<string, any> = {},
  options: { throwOn404?: boolean } = {}
): Promise<T> {
  const apiBase = getApiBase();
  const { throwOn404 = false } = options;

  try {
    const response = await axios.post(`${apiBase}${endpoint}`, params, {
      headers: { 'Content-Type': 'application/json' },
    });
    return response.data;
  } catch (error: any) {
    // Gestion gracieuse des 404 pour les endpoints CMS non implémentés
    if (error.response?.status === 404 && !throwOn404) {
      logger.warn(`Endpoint CMS non implémenté: ${endpoint}`);
      // Retourner une structure par défaut selon le type de réponse attendu
      return { success: false, error: 'Not implemented' } as T;
    }

    logger.error(`Erreur API CMS [${endpoint}]:`, error);
    const errorMessage =
      error.response?.data?.error ||
      error.response?.data?.message ||
      error.message ||
      'Unknown error';
    throw new Error(errorMessage);
  }
}

/**
 * Service CMS
 */
export const cmsService = {
  /**
   * Récupère plusieurs menus par leurs codes
   * @param codes - Liste des codes de menus (ex: ['header', 'footer_quick'])
   * @returns Map de menus indexés par code
   */
  async getMenus(codes?: string[]): Promise<Record<string, Menu>> {
    const response = await jsonrpc<{ success: boolean; menus: Record<string, Menu>; error?: string }>(
      '/menus',
      { codes }
    );
    if (response.success) {
      return response.menus;
    }
    throw new Error(response.error || 'Failed to fetch menus');
  },

  /**
   * Récupère un menu spécifique par son code
   * @param code - Code du menu (ex: 'header')
   */
  async getMenu(code: string): Promise<Menu | null> {
    const response = await jsonrpc<{ success: boolean; menu: Menu; error?: string }>(
      `/menus/${code}`,
      {},
      { throwOn404: false }
    );
    if (response.success && response.menu) {
      return response.menu;
    }
    // Retourner null si le menu n'existe pas (permet au composant de gérer gracieusement)
    return null;
  },

  /**
   * Récupère une page CMS par son slug
   * @param slug - Slug URL de la page (ex: 'about')
   */
  async getPage(slug: string): Promise<CmsPage> {
    const response = await jsonrpc<{ success: boolean; page: CmsPage; error?: string }>(
      `/pages/${slug}`,
      {}
    );
    if (response.success) {
      return response.page;
    }
    throw new Error(response.error || `Page ${slug} not found`);
  },

  /**
   * Récupère les données SEO d'une page (pour les metadata Next.js)
   * @param slug - Slug URL de la page
   */
  async getPageSeo(slug: string): Promise<PageSeo> {
    const response = await jsonrpc<{ success: boolean; seo: PageSeo; error?: string }>(
      `/pages/${slug}/seo`,
      {}
    );
    if (response.success) {
      return response.seo;
    }
    throw new Error(response.error || `SEO for ${slug} not found`);
  },

  /**
   * Liste les pages publiées avec pagination
   */
  async getPages(options?: GetPagesOptions): Promise<PagesResponse> {
    const response = await jsonrpc<{ success: boolean; pages: CmsPageSummary[]; total: number; limit: number; offset: number; error?: string }>(
      '/pages',
      options || {}
    );
    if (response.success) {
      return {
        pages: response.pages,
        total: response.total,
        limit: response.limit,
        offset: response.offset,
      };
    }
    throw new Error(response.error || 'Failed to fetch pages');
  },

  /**
   * Récupère un bloc réutilisable par son code
   * @param code - Code technique du bloc
   */
  async getBlock(code: string): Promise<CmsBlock> {
    const response = await jsonrpc<{ success: boolean; block: CmsBlock; error?: string }>(
      `/blocks/${code}`,
      {}
    );
    if (response.success) {
      return response.block;
    }
    throw new Error(response.error || `Block ${code} not found`);
  },

  /**
   * Récupère plusieurs blocs par leurs codes
   * @param codes - Liste des codes de blocs
   */
  async getBlocks(codes: string[]): Promise<Record<string, CmsBlock>> {
    const response = await jsonrpc<{ success: boolean; blocks: Record<string, CmsBlock>; error?: string }>(
      '/blocks',
      { codes }
    );
    if (response.success) {
      return response.blocks;
    }
    throw new Error(response.error || 'Failed to fetch blocks');
  },
};

export default cmsService;
